desktop/app/lib/session/index.js

```diff
@@ -0,0 +1,115 @@
+/**
+ * External Dependencies
+ */
+const EventEmitter = require( 'events' ).EventEmitter;
+
+/**
+ * Internal Dependencies
+ */
+const keychain = require( '../../lib/keychain' );
+const log = require( '../../lib/logger' )( 'desktop:session' );
+
+/*
+ * Module constants
+ */
+
+class SessionManager extends EventEmitter {
+	constructor() {
+		super();
+	}
+
+	async init( window ) {
+		log.info( 'Initializing session manager...' );
+
+		this.loggedIn = false;
+		this.window = window;
+
+		// Check for existing cookies
+		const wordpress_logged_in = await getCookie(
+			window,
+			'https://public-api.wordpress.com',
+			'wordpress_logged_in'
+		);
+		if ( wordpress_logged_in && ! this.loggedIn ) {
+			log.info( `Got 'wordpress_logged_in' cookie, emitting 'logged-in' event...` );
+
+			this.loggedIn = true;
+			this.emit( 'logged-in', wordpress_logged_in );
+			log.debug( `Logged in with cookie 'wordpress_logged_in': `, wordpress_logged_in );
+		}
+
+		const wp_api_sec = await getCookie( window, 'https://public-api.wordpress.com', 'wp_api_sec' );
+		if ( this.loggedIn && wp_api_sec ) {
+			await keychainWrite( 'wp_api_sec', wp_api_sec.value );
+			this.emit( 'api:connect' );
+		}
+
+		// Listen for auth events
+		this.window.webContents.session.cookies.on(
```

Comment:
That is some deep chaining! Any chance any of this could be undefined?

Reply:
Based on the docs it looks like these objects always exist on the `BrowserWindow` object. My understanding is that `webContents`, `session` and `cookies` are permanent objects on the `BrowserWindow` that provide event listeners for various events (i.e. they exist outside the life cycle of the events happening).